[AAP-48398] Add periodic role assignment fallback sync

jessicamack opened the pull request:
## Description
<!-- Mandatory: Provide a clear, concise description of the changes and their purpose -->
Gateway is the source of truth for RBAC. This change adds a method to sync the Role Team and Role User assignments between Controller and Gateway using the resource server. This ensures that all Role assignments locally match what Gateway reports.

## Type of Change
<!-- Mandatory: Check one or more boxes that apply -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist
<!-- These items help ensure quality - they complement our automated CI checks -->
- [X] I have performed a self-review of my code
- [X] I have added relevant comments to complex code sections
- [] I have updated documentation where needed
- [X] I have considered the security impact of these changes
- [X] I have considered performance implications
- [X] I have thought about error handling and edge cases
- [ ] I have tested the changes in my local environment

## Testing Instructions
<!-- Optional for test-only changes. Mandatory for all other changes -->
<!-- Must be detailed enough for reviewers to reproduce -->
### Prerequisites
<!-- List any specific setup required -->

### Steps to Test
1. 
2. 
3. 

### Expected Results
<!-- Describe what should happen after following the steps -->

## Additional Context
<!-- Optional but helpful information -->

### Required Actions
<!-- Check if changes require work in other areas -->
<!-- Remove section if no external actions needed -->
- [ ] Requires documentation updates
  <!-- API docs, feature docs, deployment guides -->
- [ ] Requires downstream repository changes
  <!-- Specify repos: django-ansible-base, eda-server, etc. -->
- [ ] Requires infrastructure/deployment changes

### Screenshots/Logs
<!-- Add if relevant to demonstrate the changes -->

jessicamack committed:
parent a22e13950be108274a376edfcd38fe7128e07efe
author jessicamack <jmack@redhat.com> 1755536606 -0400
committer jessicamack <jmack@redhat.com> 1755633723 -0400

parent a22e13950be108274a376edfcd38fe7128e07efe
author jessicamack <jmack@redhat.com> 1755536606 -0400
committer jessicamack <jmack@redhat.com> 1755633623 -0400

add role assignment sync and unit test

lint fixes

lint fix

fix mock

change mock

reformat test

fix test

update test

reformat test

add role assignment sync and unit test

lint fixes

lint fix

change mock

reformat test

fix test

lint fix

update test

lint fix

fix test. remove unneeded file

AlanCoding committed:
AAP-50805 Ignore role content types not known to system in sync (#811)

The bug here is that AWX `periodic_resource_sync` will result in an
error.

It seems there was an attempt to log an error and continue on when
getting a `ValidationError`, it seems this only targeted the Django
validation error exception, leaving the more common DRF ValidationError
uncaught. This means that the issue broke the resource sync, globally.
That means the periodic sync was effectively broken by the referenced
bug.

We could change that, but we wanted to have a targeted and more
intentional special-case just for roles, and that is what this
implements.

If the field `RoleDefinition.content_type` references a type that does
not exist locally, we don't bother in the _periodic fallback_ sync and
just skip over that record. This is very sensible behavior, and it will
be reported as a no-op, as one should expect.

jessicamack committed:
Merge branch 'devel' into roleassignment_fallback_sync

jessicamack committed:
fix lint issues and remove duplication from rebase

fosterseth commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +137,193 @@ def fetch_manifest(
+        # Get the object if it's not a global assignment
+        content_object = None
+        if assignment_tuple.object_id:
+            object_resource = Resource.objects.get(ansible_id=assignment_tuple.object_id)
```
currently in `get_remote_assignments` object_id can be ansible id or just pk
```
object_id = assignment.get('object_id') or assignment.get('object_ansible_id')
```
but then we do a lookup here with `ansible_id=` which won't be correct if the object_id is not an ansible_id

ultimately I am seeing this error
```
ValidationError(['“1” is not a valid UUID.'])
```

jessicamack committed:
Merge branch 'devel' into roleassignment_fallback_sync

jessicamack committed:
apply feedback

change tuple field name to be more descriptive,
order of checking ids,
look for model instead of Resource if id is not ansible_id

fosterseth commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +137,196 @@ def fetch_manifest(
+                model = role_definition.content_type.model_class()
+                content_object = model.objects.get(pk=assignment_tuple.ansible_id_or_pk)
+
+            if assignment_tuple.assignment_type == 'user':
```
this code block seems unnecessary to me, as all delete cases should be taken care of in
```
role_definition.remove_permission(
```

jessicamack committed:
remove redundant code

also catch missed uses of old field name

jessicamack committed:
remove unused imports. rename field in test

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -458,6 +664,59 @@ def _dispatch_sync_process(self, manifest_list: list[ManifestItem]):
+            # Store results for reporting
+            self.results["assignments_created"] = created_count
+            self.results["assignments_deleted"] = deleted_count
+            self.results["assignment_errors"] = error_count
```
This data type is unexpected by AWX

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[9], line 1
----> 1 periodic_resource_sync()

File /src/controller/awx/main/tasks/system.py:1034, in periodic_resource_sync()
   1032     continue
   1033 # Log creations and conflicts
-> 1034 if len(item_list) > 10 and settings.LOG_AGGREGATOR_LEVEL != 'DEBUG':
   1035     logger.info(f'Periodic resource sync {key}, first 10 items:\n{item_list[:10]}')
   1036 else:

TypeError: object of type 'int' has no len()
```

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +137,188 @@ def fetch_manifest(
+                try:
+                    object_resource = Resource.objects.filter(object_id=object_id, content_type=assignment.content_type).first()
+                    if object_resource:
+                        object_id = object_resource.ansible_id
```
probably need to `str()` this

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +137,188 @@ def fetch_manifest(
+            object_id = assignment.object_id
+            if object_id and assignment.content_type:
+                # For object-scoped assignments, try to get the object's ansible_id
+                try:
```
It would be better to use the same logic, like checking the model being 'organization' or 'team'. And dump that into a static method for better consolidation.

jessicamack committed:
lint fix. make results a list

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +137,188 @@ def fetch_manifest(
+            if object_id and assignment.content_type:
+                # For object-scoped assignments, try to get the object's ansible_id
+                try:
+                    object_resource = Resource.objects.filter(object_id=object_id, content_type=assignment.content_type).first()
```
`content_type=assignment.content_type` that won't be the same model, probably throwing error

jessicamack committed:
apply feedback

removed some try/excepts,
moved imports,
added helper methods,
renamed variables for clarity

jessicamack committed:
lint fix. update test to match changes

jessicamack committed:
fix bug

"cannot access local variable 'content_object' where it is not associated with a value"

fosterseth commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -73,6 +75,29 @@ def __iter__(self):
+    actor_ansible_id: str  # user_ansible_id or team_ansible_id
+    ansible_id_or_pk: str | None  # object_id or object_ansible_id (None for global)
+    role_definition_name: str
+    assignment_type: str  # 'user' or 'team'
```
seems this is no longer being used at all, if so we can drop it

jessicamack committed:
add local variable

fosterseth reviewed (approved):
tested it manually by creating resources in gateway and exec-ing into awx, deleting assignments, and running periodic_resource_sync. I can see that it recreates the assignments as intended, deletes when necessary. Handles both global and non-global roles.

Also re-running command back to back is no-op on subsequent runs.

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +139,185 @@ def fetch_manifest(
+
+    # Fetch user assignments
+    try:
+        user_resp = api_client.list_user_assignments()
```
This isn't paginated. That's going to be a blocker.

there is a kwarg you can use to pass query params, for which it would be best to do a 200 page size, but even then you need a loop to go over all the pages.

fosterseth replied:
implemented pagination

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -114,6 +139,185 @@ def fetch_manifest(
+            )
+        )
+
+    return assignments
```
For memory use an isolation I think it would clearly be _better_ do do the set comparisons for user & team assignments separately, so you are not building a set in memory of both combined. But I realize this probably isn't going to happen before merge of this.

fosterseth replied:
agreed this can get pretty taxing as it is written, but we can plan to do this later

AlanCoding commented on ansible_base/resource_registry/tasks/sync.py:
```diff
@@ -458,6 +663,59 @@ def _dispatch_sync_process(self, manifest_list: list[ManifestItem]):
+
+        except Exception as e:
+            self.write(f"Assignment sync failed: {e}")
+            logger.error(f"Assignment sync failed: {e}")
```
I would prefer to get more information, or else we're likely to cause unnecessary iteration in debugging this

```suggestion
            logger.exception(f"Assignment sync failed")
```

jessicamack committed:
log exception instead of error

Co-authored-by: Alan Rominger <arominge@redhat.com>

fosterseth committed:
Add pagination filter when requesting from gateway

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

jessicamack committed:
lint fix

fosterseth committed:
just debug level

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth commented:
tested that pagination works by creating 400 assignments in gateway. Here are the logs
```
>>> periodic_resource_sync()
2025-08-21 17:22:10,322 DEBUG    [-] awx.main.tasks.system Running periodic resource sync
2025-08-21 17:22:13,441 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 2 of user assignments
2025-08-21 17:22:13,566 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 3 of user assignments
2025-08-21 17:22:13,704 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 4 of user assignments
2025-08-21 17:22:13,820 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 5 of user assignments
2025-08-21 17:22:13,943 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 6 of user assignments
2025-08-21 17:22:14,051 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 7 of user assignments
2025-08-21 17:22:14,153 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 8 of user assignments
2025-08-21 17:22:14,249 WARNING  [-] ansible_base.resources_api.tasks.sync Fetching next page 9 of user assignments
2025-08-21 17:22:35,609 INFO     [-] awx.main.tasks.system Periodic resource sync assignments_created:
[401]
2025-08-21 17:22:35,610 INFO     [-] awx.main.tasks.system Periodic resource sync assignments_deleted:
[0]
2025-08-21 17:22:35,611 INFO     [-] awx.main.tasks.system Periodic resource sync assignment_errors:
[0]
```

fosterseth committed:
Merge branch 'devel' into roleassignment_fallback_sync

jlmitch5 commented:
After discussion, we have decided that the tests are well-covered.  On attempt to improve the coverage asked for, it would create a lot of noisy tests that serve no effective purpose.  As such we are going to bypass this check.  If there is a problem with this, please reach out to me
